libclc/ptx-nvidiacl/libspirv/group/collectives.cl

```diff
@@ -0,0 +1,419 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <spirv/spirv.h>
+#include <spirv/spirv_types.h>
+
+#pragma OPENCL EXTENSION cl_khr_fp16 : enable
+#pragma OPENCL EXTENSION cl_khr_fp64 : enable
+
+// CLC helpers
+__local bool *
+__clc__get_group_scratch_bool() __asm("__clc__get_group_scratch_bool");
+__local char *
+__clc__get_group_scratch_char() __asm("__clc__get_group_scratch_char");
+__local uchar *
+__clc__get_group_scratch_uchar() __asm("__clc__get_group_scratch_char");
+__local short *
+__clc__get_group_scratch_short() __asm("__clc__get_group_scratch_short");
+__local ushort *
+__clc__get_group_scratch_ushort() __asm("__clc__get_group_scratch_short");
+__local int *
+__clc__get_group_scratch_int() __asm("__clc__get_group_scratch_int");
+__local uint *
+__clc__get_group_scratch_uint() __asm("__clc__get_group_scratch_int");
+__local long *
+__clc__get_group_scratch_long() __asm("__clc__get_group_scratch_long");
+__local ulong *
+__clc__get_group_scratch_ulong() __asm("__clc__get_group_scratch_long");
+__local half *
+__clc__get_group_scratch_half() __asm("__clc__get_group_scratch_half");
+__local float *
+__clc__get_group_scratch_float() __asm("__clc__get_group_scratch_float");
+__local double *
+__clc__get_group_scratch_double() __asm("__clc__get_group_scratch_double");
+
+_CLC_DEF _CLC_CONVERGENT uint __clc__membermask() {
+  uint FULL_MASK = 0xFFFFFFFF;
+  uint max_size = __spirv_SubgroupMaxSize();
+  uint sg_size = __spirv_SubgroupSize();
+  return FULL_MASK >> (max_size - sg_size);
+}
+
+#define __CLC_SUBGROUP_SHUFFLE_I32(TYPE)                                       \
+  _CLC_DEF _CLC_OVERLOAD _CLC_CONVERGENT TYPE __clc__SubgroupShuffle(          \
+      TYPE x, uint idx) {                                                      \
+    return __nvvm_shfl_sync_idx_i32(__clc__membermask(), x, idx, 0x1f);        \
+  }
+__CLC_SUBGROUP_SHUFFLE_I32(char);
+__CLC_SUBGROUP_SHUFFLE_I32(uchar);
+__CLC_SUBGROUP_SHUFFLE_I32(short);
+__CLC_SUBGROUP_SHUFFLE_I32(ushort);
+__CLC_SUBGROUP_SHUFFLE_I32(int);
+__CLC_SUBGROUP_SHUFFLE_I32(uint);
+#undef __CLC_SUBGROUP_SHUFFLE_I32
+
+_CLC_DEF _CLC_OVERLOAD _CLC_CONVERGENT ulong __clc__SubgroupShuffle(ulong x,
+                                                                    uint idx) {
+  uint lo = (uint)x;
+  uint hi = (uint)(x >> 32);
+  lo = __nvvm_shfl_sync_idx_i32(__clc__membermask(), lo, idx, 0x1f);
+  hi = __nvvm_shfl_sync_idx_i32(__clc__membermask(), hi, idx, 0x1f);
+  return (((ulong)hi) << 32ul) | lo;
```

Comment:
You might consider casting `ulong` to `uint2` to improve readability.
```suggestion
  uint2 y = as_uint2(x);
  y.lo = __nvvm_shfl_sync_idx_i32(__clc__membermask(), y.lo, idx, 0x1f);
  y.hi = __nvvm_shfl_sync_idx_i32(__clc__membermask(), y.hi, idx, 0x1f);
  return as_ulong(y);
```

The same applies to `__clc__SubgroupShuffleUp`.

Reply:
Ooh, that's much nicer.  Thanks.  Fixed in 3b044955bb9994a3ea6e3dc5efbbee9a6bdc5f24.